packages/headless/src/ssr-next/commerce/utils/controller-wiring.ts

```diff
@@ -1,18 +1,195 @@
+  }),
+};
+
+const listingDefinition = {
+  ...requiredDefinition,
+  searchParams: new RecordValue({
+    values: parametersDefinition,
+  }),
+};
+
+const standaloneDefinition = {
+  ...requiredDefinition,
+};
+
+const searchDefinition = {
+  ...requiredDefinition,
+  searchParams: new RecordValue({
+    options: {required: true},
+    values: {query: requiredEmptyAllowedString, ...parametersDefinition},
+  }),
+};
+
+const recommendationsDefinition = {
+  ...requiredDefinition,
+  // recommendation:  // TODO: KIT-4619: support array of recommendations
+};
+
+export const listingDefinitionSchema = new Schema(listingDefinition);
+export const searchDefinitionSchema = new Schema(searchDefinition);
+export const standaloneDefinitionSchema = new Schema(standaloneDefinition);
+export const recommendationsDefinitionSchema = new Schema(
+  recommendationsDefinition
+);
+
+/**
+ * Validates the build configuration based on the solution type.
+ */
+function validateBuildConfig(
+  solutionType: SolutionType,
+  buildConfig: BuildConfig<SolutionType>
+): void {
+  const validationMap: Record<SolutionType, Schema<object>> = {
+    [SolutionType.listing]: listingDefinitionSchema,
+    [SolutionType.search]: searchDefinitionSchema,
+    [SolutionType.standalone]: standaloneDefinitionSchema,
+    [SolutionType.recommendation]: recommendationsDefinitionSchema,
+  };
+
+  const schema = validationMap[solutionType as SolutionType];
+  schema.validate(buildConfig);
+}
+
+function createControllerWirer(
+  buildConfig: BuildConfig<SolutionType>,
+  controllerDefinitions: CommerceControllerDefinitionsMap,
+  controllerProps: ControllersPropsMap
+) {
+  return {
+    wireParameterManager: () => {
+      if (!controllerDefinitions?.parameterManager) return;
+
+      const {searchParams} = buildConfig;
+      const {query, ...rest} =
+        (searchParams as Parameters & {
+          query?: string;
+        }) || {};
+
+      const parameters = {
+        ...(query && {q: query}),
+        ...(rest && typeof rest === 'object' ? rest : {}),
+      };
+
+      controllerProps.parameterManager = {
+        initialState: {parameters},
+      };
+    },
+
+    wireContext: () => {
+      if (!controllerDefinitions?.context) return;
+
+      const {context} = buildConfig;
+      controllerProps.context = {
+        initialState: {
+          ...context,
+        },
+      };
+    },
+
+    wireCart: () => {
+      if (!controllerDefinitions?.cart || !buildConfig.cart) return;
+
+      controllerProps.cart = {
+        initialState: buildConfig.cart,
+      };
+    },
+
+    wireRecommendations: () => {
+      // TODO: KIT-4619: wire recommendations
+    },
+  };
+}
+
+function wireCommonControllers(
+  wirer: ReturnType<typeof createControllerWirer>
+): void {
+  wirer.wireCart();
+  wirer.wireContext();
+}
+
+function wireSolutionSpecificControllers(
+  solutionType: SolutionType,
+  wirer: ReturnType<typeof createControllerWirer>
+): void {
+  switch (solutionType) {
```

Comment:
I think another way (maybe simpler? ill let you decide...) would be to have a single single `wire` function ?

so like:
```ts
wire(solutionType, buildConfig, controllerDefinitions, controllerProps)
```

or
```ts
new Wirer(solutionType, buildConfig, controllerDefinitions, controllerProps).wire()
```
if you prefer a class.

That does the full job in a sequential manner. And can still split each `wireXYZ` function internally for readability.